### PR TITLE
fix(ngcc): support element accesses for export declarations

### LINKED ISSUE #44037: ngcc: processing of recently published UMD bundles may fail
# Problem description

Recently, a number of issues (#43870, #43936, #44019) has started to occur for certain UMD bundles when ngcc (the Angular Compatibility Compiler) processes them to create an Ivy-compatible version of the UMD bundle.

Recent changes to how UMD bundles are formatted are not supported by the current Angular Compatibility Compiler (ngcc), which can result in a fatal error, like the following:

- `TypeError: Cannot read property 'text' of undefined`
- `Error: Failed to find exported name of node`

Packages from the `@angular` scope are not affected, but recent versions of third-party libraries may result in one of the above fatal errors.

In the typical Angular CLI setup UMD bundles are not generally needed, as ES2015 bundles are preferred over UMD. It is therefore expected that CLI users won't be affected by these issues.

# Available workarounds

If ngcc is manually executed from outside of the CLI, without the `--properties` argument, then it will process all available formats in a library including the UMD bundles. In that case, ngcc may fail if a library is used that contains the newly structured code that ngcc is currently unable to understand. It may be that the UMD bundles are not actually necessary, in which case ngcc can be configured not to process the UMD bundles at all. To this effect, pass the `--properties es2015 module` argument to ngcc.

If you do require the UMD bundles to be processed, for example if you're using Jest for unit testing, then the only workaround that is currently available is to pin the dependency that breaks ngcc to an earlier version which does not contain the problematic code patterns that cause ngcc to fail.

# Next steps

Although some of the changes have already been accounted for in the latest Angular 12 releases, we're still working on a solution to resolve all ngcc failures of UMD bundles.

### PR DESCRIPTION
**Pull request by JoostK**
Bundlers like Rollup may use an element access expression for an export
declaration, which causes ngcc to ignore those export declarations possibly
resulting in incomplete processing of packages.

Element access syntax may be used when the declared name is not considered
as valid JS identifier, but bundlers may be conservative in determining whether
an identifier can be used (to emit a property access) and opt for a string
literal in an element access instead.

The element access syntax introduces a problem for ngcc, where it wouldn't
consider such export as class declaration, causing them to be skipped. The
ngtsc compiler is implemented with the assumption that all class declarations
use a `ts.Identifier` as name, whereas the element access is using a string
literal for the declared name. This makes it troublesome for ngcc to support
this syntax form in UMD bundles.

To work around the problem, this function transforms these access expressions
into regular property accesses. The source text is parsed to an AST to allow
finding the element accesses in a robust way, after which the affected text
ranges are replaced with property accesses in the original source text.

Closes #44037

### CODE REVIEW

**Review by gkalpak — APPROVED** [review]
LGTM 👏

Do we know how much this affects processing time for UMDs?

File: packages/compiler-cli/ngcc/src/packages/adjust_cjs_exports.ts [hunk 1]
```diff
@@ -0,0 +1,82 @@
+/**
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/ngcc/src/packages/adjust_cjs_exports.ts:1]: It wasn't a deliberate omission; I've renamed to file to include UMD.

**Comment by JoostK** [comment]
> Do we know how much this affects processing time for UMDs?

I haven't checked but I suspect it should be negligible with the short-circuiting check. I'll run it through ngcc-validation to verify (mostly to ensure that it doesn't break anything, I don't think any perf difference can be observed on CI).

**Comment by JoostK** [comment]
Seems to have gone well: https://github.com/angular/ngcc-validation/pull/5270

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit b66e479cdb1e474a29ff676f10a5fcc3d7eae799.

**Pull request closed without merging**